### PR TITLE
dhcp: T3316: Adjust kea lease files' location and permissions

### PR DESCRIPTION
**Pull request by indrajitr**
## Change Summary

Move the kea lease file to a separate directory `/config/dhcp` that `kea` process can write to so that subprocesses spawned by `kea` process can operate on the lease files.
 
To allow `kea` process to write to `/config/dhcp`, add `_kea` user to `vyattacfg` group. And the lease files are owned completely by `_kea` user to play well with `kea-lfc` process.
 
Specifically, this is necessary for `kea-lfc` which is spawned by `kea` process to clean up expired leases. Since `kea-lfc` creates additional backup lease files, it needs write access to the lease file directory.

Additionally, change the extension of the lease file from `.leases` to `.csv` to reflect the actual file format.

See:
* https://gitlab.isc.org/isc-projects/kea/-/issues/833
* https://lists.isc.org/pipermail/kea-users/2016-April/000328.html

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)

* https://vyos.dev/T3316

## Component(s) name

dhcp server, dhcpv6 server

## How to test

1. Check that `_kea` user is a member of `vyattacfg`:
```
vyos@v15-c28:~$ getent group vyattacfg | grep '_kea'
vyattacfg:x:102:tacacs15,radius_priv_user,sysadmin,_kea
```

2. Configure `service dhcp-server` as usual

3. Look for the directories, files and their permission in `/config`:
   - `/config/dhcp` should be writable by `vyattacfg` group
   - `/config/dhcp/*` should be owned by `_kea`
```
vyos@v15-c28:~$ ls -ld /config/dhcp
drwxrwxr-x 2 root vyattacfg 4096 Dec 29 00:20 /config/dhcp

vyos@v15-c28:~$ ls -ls /config/dhcp
total 12
4 -rw-r--r-- 1 _kea _kea      103 Dec 29 00:16 dhcp4-leases.csv
4 -rw-r--r-- 1 _kea _kea      462 Dec 29 00:16 dhcp4-leases.csv.2
4 -rw-r--r-- 1 _kea _kea      160 Dec 29 00:20 dhcp6-leases.csv
```

## Smoketest result

```
vyos@v15-c28:~$ /usr/libexec/vyos/tests/smoke/cli/test_service_dhcp-server.py 
test_dhcp_exclude_in_range (__main__.TestServiceDHCPServer.test_dhcp_exclude_in_range) ... ok
test_dhcp_exclude_not_in_range (__main__.TestServiceDHCPServer.test_dhcp_exclude_not_in_range) ... ok
test_dhcp_failover (__main__.TestServiceDHCPServer.test_dhcp_failover) ... 
No DHCP address range or active static-mapping configured within shared-
network "FAILOVER, 192.0.2.0/25"!

ok
test_dhcp_multiple_pools (__main__.TestServiceDHCPServer.test_dhcp_multiple_pools) ... ok
test_dhcp_relay_server (__main__.TestServiceDHCPServer.test_dhcp_relay_server) ... ok
test_dhcp_single_pool_options (__main__.TestServiceDHCPServer.test_dhcp_single_pool_options) ... 
No DHCP address range or active static-mapping configured within shared-
network "SMOKE-0815, 192.0.2.0/25"!

ok
test_dhcp_single_pool_range (__main__.TestServiceDHCPServer.test_dhcp_single_pool_range) ... 
No DHCP address range or active static-mapping configured within shared-
network "SMOKE-1, 192.0.2.0/25"!

ok
test_dhcp_single_pool_static_mapping (__main__.TestServiceDHCPServer.test_dhcp_single_pool_static_mapping) ... 
No DHCP address range or active static-mapping configured within shared-
network "SMOKE-2, 192.0.2.0/25"!


Either MAC address or Client identifier (DUID) is required for static
mapping "client1" within shared-network "SMOKE-2, 192.0.2.0/25"!

ok

----------------------------------------------------------------------
Ran 8 tests in 203.195s

OK
vyos@v15-c28:~$ /usr/libexec/vyos/tests/smoke/cli/test_service_dhcpv6-server.py 
test_global_nameserver (__main__.TestServiceDHCPv6Server.test_global_nameserver) ... ok
test_prefix_delegation (__main__.TestServiceDHCPv6Server.test_prefix_delegation) ... ok
test_single_pool (__main__.TestServiceDHCPv6Server.test_single_pool) ... 
Either MAC address or Client identifier (DUID) is required for static
mapping "client1" within shared-network "SMOKE-1, 2001:db8:f00::/64"!

ok

----------------------------------------------------------------------
Ran 3 tests in 84.187s

OK
```

## Checklist:

- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [x] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

**Review by c-po — APPROVED** [review]

**Comment by indrajitr** [comment]
TL;DR for reviewers:
1. Leases go in `/config/dhcp` and not in top level `/config`
2. `/config/dhcp` is owned by `vyattacfg` as usual
3. `_kea` is a member of `vyattacfg` group
4. Kea leases go under `/config/dhcp` and does not clutter top level `/config`
5. Kea process (and thus its subprocess `kea-lfc`) has unfettered access in `/config/dhcp`

Open question:
I named the lease directory as `/config/dhcp`. But there could be other options (like `/config/leases`, `config/kea` etc.). I am open to suggestions if a different name is preferred)

Hope I'm not missing anything :)

File: src/conf_mode/system-login.py [hunk 1]
```diff
@@ -331,7 +331,7 @@ def apply(login):
             if tmp: command += f" --home '{tmp}'"
             else: command += f" --home '/home/{user}'"
 
-            command += f' --groups frr,frrvty,vyattacfg,sudo,adm,dip,disk,_kea {user}'
```
Comment (sarthurdev) [inline src/conf_mode/system-login.py]: Iirc this was necessary to access the control socket from op-mode scripts.

Comment (indrajitr) [reply src/conf_mode/system-login.py]: Right you are! I can see op-mode commands will need this. Thanks!

Comment (indrajitr) [reply src/conf_mode/system-login.py]: Interestingly, op-mode commands still fail. Something is going on.

Here is what I have:

- User `vyos` belongs to `_kea` group:
```
vyos@v15-c28:~$ getent group _kea
_kea:x:114:vyos
vyos@v15-c28:~$ grep _kea /etc/group
vyattacfg:x:102:tacacs15,radius_priv_user,vyos,_kea
_kea:x:114:vyos
```

- ctl-socket has 775:
```
vyos@v15-c28:~$ ls -l /run/kea/dhcp4-ctrl-socket
srwxrwxr-x 1 _kea _kea 0 Dec 29 02:56 /run/kea/dhcp4-ctrl-socket
```

- And still I see:
```
vyos@v15-c28:~$ show dhcp server leases 
Traceback (most recent call last):
  File "/usr/libexec/vyos/op_mode/dhcp.py", line 394, in <module>
    res = vyos.opmode.run(sys.modules[__name__])
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/vyos/opmode.py", line 263, in run
    res = func(**args)
          ^^^^^^^^^^^^
  File "/usr/libexec/vyos/op_mode/dhcp.py", line 261, in _wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/libexec/vyos/op_mode/dhcp.py", line 294, in show_server_leases
    lease_data = _get_raw_server_leases(family=family, pool=pool, sorted=sorted, state=state, origin=origin)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/libexec/vyos/op_mode/dhcp.py", line 88, in _get_raw_server_leases
    active_config = kea_get_active_config(inet_suffix)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/vyos/kea.py", line 318, in kea_get_active_config
    config = _ctrl_socket_command(ctrl_socket, 'config-get')
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/vyos/kea.py", line 299, in _ctrl_socket_command
    sock.connect(path)
PermissionError: [Errno 13] Permission denied
```

Comment (indrajitr) [reply src/conf_mode/system-login.py]: A clean build solved for it.

File: debian/vyos-1x.postinst [hunk 2]
```diff
@@ -109,11 +109,10 @@ if ! grep -q '^hostsd' /etc/group; then
...
 
-# add dhcpd user for dhcp-server
-if ! grep -q '^dhcpd' /etc/passwd; then
-    adduser --quiet --system --disabled-login --no-create-home --home /run/dhcp-server dhcpd
-    adduser --quiet dhcpd hostsd
-fi
+# Add _kea user for kea-dhcp{4,6}-server to vyattacfg
+# Note that kea-common should have been installed as dependency and the user
+# '_kea' should exist
+adduser --quiet _kea vyattacfg
```
Comment (c-po) [inline debian/vyos-1x.postinst]: Please add a test if user _kea exists.

```bash
if  grep -q '^_kea' /etc/passwd; then
    adduser --quiet _kea vyattacfg
fi
```

**Review by c-po — CHANGES_REQUESTED** [review]

**Comment by c-po** [comment]
I like the tl;dr - /config/dhcp is fine as we try to not hardcode names from processes/software and rather use generic terms

File: debian/vyos-1x.postinst (hunk 2 above)
Comment (sarthurdev) [reply debian/vyos-1x.postinst]: If the user exists, should it use `usermod` to still add the group?

Comment (c-po) [reply debian/vyos-1x.postinst]: `adduser` is fine doing this

```
cpo@LR1.wue3# sudo adduser _kea vyattacfg
Adding user `_kea' to group `vyattacfg' ...
Done.
```

File: debian/vyos-1x.postinst [hunk 3]
```diff
@@ -109,10 +109,10 @@ if ! grep -q '^hostsd' /etc/group; then
     addgroup --quiet --system hostsd
 fi
 
-# add dhcpd user for dhcp-server
-if ! grep -q '^dhcpd' /etc/passwd; then
-    adduser --quiet --system --disabled-login --no-create-home --home /run/dhcp-server dhcpd
-    adduser --quiet dhcpd hostsd
```
Comment (sarthurdev) [inline debian/vyos-1x.postinst:115]: `hostsd` group likely still required on `_kea` for the hook to update hosts file via vyos-hostsd service.

**Review by sarthurdev — CHANGES_REQUESTED** [review]

File: debian/vyos-1x.postinst (hunk 3 above)
Comment (indrajitr) [reply debian/vyos-1x.postinst:115]: As it stands currently, kea probably doesn't operate as the user `dhcpd`. So this looked dead-code to me.

That said, I wonder under what circumstances `_kea` would require `hostsd` group. Based on what I am seeing, `/usr/libexec/vyos/system/on-dhcp-event.sh` (executed as `_kea`) is the script that invokes `vyos-hostsd-client` and dispatches relevant JSON payload to `vyos-hostsd` (the server). Unless `vyos-hostd` imposes restriction on the client uid/gid, we'll probably be okay.

Comment (indrajitr) [reply debian/vyos-1x.postinst:115]: Update: Just deployed this branch "as-is" on my router, I can see DHCP leases are populating `/etc/hosts`. So `vyos-hostsd` is honoring the messages from `vyos-hostsd-client` that is running as `_kea` user it seems.

**Review by c-po — APPROVED** [review]

**Comment by 24fpsDaVinci** [comment]
i built my own iso this morning with this patch and during migration to https://github.com/vyos/vyos-rolling-nightly-builds/releases/tag/1.5-rolling-202401010026 the files in /config/dhcp/ changed ownership to root from _kea

**Comment by c-po** [comment]
This is addressed in https://github.com/vyos/vyos-1x/pull/2731
